ivadomed/loader/loader.py

```diff
@@ -1121,3 +1053,91 @@ def __init__(self, bids_df, subject_file_lst, target_suffix, contrast_params, sl
+        multichannel_subjects[subj_id]["metadata"][idx] = metadata
+        if roi_filename:
+            multichannel_subjects[subj_id]["roi_filename"] = roi_filename
+        return multichannel_subjects
+
+
+    def create_filename_pair(self, multichannel_subjects, subject, c, tot, multichannel, df_subjects, contrast_params, 
+                            target_suffix, all_deriv, bids_df, bounding_box_dict, idx_dict, metadata_choice):
+        df_sub = df_subjects.loc[df_subjects['filename'] == subject]
+
+        # Training & Validation: do not consider the contrasts over the threshold contained in contrast_balance
+        contrast = df_sub['suffix'].values[0]
+        if contrast in (contrast_params["balance"].keys()):
+            c[contrast] = c[contrast] + 1
+            if c[contrast] / tot[contrast] > contrast_params["balance"][contrast]:
+                return
+        if isinstance(target_suffix[0], str):
+            target_filename, roi_filename = [None] * len(target_suffix), None
+        else:
+            target_filename, roi_filename = [[] for _ in range(len(target_suffix))], None
+
+        derivatives = bids_df.df[bids_df.df['filename']
+                        .str.contains('|'.join(bids_df.get_derivatives(subject, all_deriv)))]['path'].to_list()
+
+        for derivative in derivatives:
+            target_filename = self.annotate_raters(target_suffix, target_filename, derivative)
+            if not (self.roi_params["suffix"] is None) and self.roi_params["suffix"] in derivative:
+                roi_filename = [derivative]
+
+        if (not any(target_filename)) or (not (self.roi_params["suffix"] is None) and (roi_filename is None)):
+            return
+
+        metadata = df_sub.to_dict(orient='records')[0]
+        metadata['contrast'] = contrast
+
+        if len(bounding_box_dict):
+            # Take only one bounding box for cropping
+            metadata['bounding_box'] = bounding_box_dict[str(df_sub['path'].values[0])][0]
+
+        if metadata_choice == 'mri_params':
+            if not all([imed_film.check_isMRIparam(m, metadata, subject, self.metadata) for m in
+                        self.metadata.keys()]):
+                return
+
+        elif metadata_choice and metadata_choice != 'contrasts' and metadata_choice is not None:
+            self.create_metadata_dict(metadata_choice, df_sub, bids_df)
+        
+        # Fill multichannel dictionary
+        # subj_id is the filename without modality suffix and extension
+        if multichannel:
+            multichannel_subjects = self.fill_multichannel_dict(multichannel_subjects, subject, idx_dict, df_sub, 
+                                                                roi_filename, target_filename, metadata)
+        else:
+            self.filename_pairs.append(([df_sub['path'].values[0]],
+                                        target_filename, roi_filename, [metadata]))
+        return multichannel_subjects
```

Comment:
Maybe the function should not have a return, and these lines could remain in the for loop outside of `create_filename_pair`. It would be clearer that the behavior is different when multichannel or not.